list.py

```diff
@@ -117,15 +117,16 @@ def __setitem__(self, index, elem):
         post: 
             sets element by index
         '''
-        index = self.check_index_and_transform(index)
+        self.check_index(index)
+
         self.find_node(index).elem = elem
         
 
     def __iter__(self):
         return ListIterator(self.head)
 
 
-    def check_index_and_transform(self, index, include_border = 0):
+    def check_index(self, index, include_border = 0):
```

Comment:
Can I say that `check_index` is an internal List method that users don't have access to?

Reply:
I think so, yes. I've never seen such method being public on any collections. In general, when deciding whether something should be considered public you may follow the following thought process:

1. Will the class still be usable according to its original purpose and design, if the member (field or method) is private? **If yes, keep the member private**
2. Is there a scenario, where a problem can not be solved without having the member public? **If no, keep the member private**
3. Is it possible that if we make the member public, then the users of the class may violate some of its internal invariants? **If yes, then keep the member private. If not still consider keeping it private if criteria above don't require it to be public**
  The term "invariants" here means - the assumptions that we must make sure always uphold for the class. An example of such an assumption would be the folowing: "The `length` field is always a non-negative number, that is equal to the number of nodes in the list". If we decide to make the `length` field public, then the users of the class may overwrite it with a negative number, or some other number which isn't equal to the number of nodes. It means we must keep this field private.

An example from real life would be:

Imagine you have a gas meter installed in your flat, which displays the volume of gas you've consumed. It is used to calculate how much money you need to pay each month for communal services.

If the gas meter were to have a button for you that you could press and it would reset the gas meter to zero, then it means the gas meter is unusable, because according to its internal invariants the volume of gas that you consumed may only grow, and never go back to zero, so that the communal services could trust these measurements. This is why gas meters usually have a "factory seal" that when damaged means that you tried to access the private guts of the gas meter potentially breaking its invariants.

This is a lot of talking for just `check_index` method. But the 3 criteria above say this method should be private. In general, if you don't know if something should be public or not, just make it private. Privacy by default is the feature of Rust and is the generally agreed practice that all new programming languages and tools follow.

If you made something mistakenly private - then it is less harm, because you just restricted your class and some user was unsatisfied because they couldn't do something, and you'll just quickly release a new version of your library and make the member public to make the user happy.

If you made something mistakenly public - then it is a deal breaker and potential end of the world, because you can't easily make public things private. The users of your library may already use your public member in a lot of places and heavily depend on them.

See it like the problem of giving a candy to the child: if you never gave a candy (public member) to a child, then the child is slightly upset, but not that much because you never promised them a candy, so they didn't expect a candy from you.

If you gave a candy to a child from day zero, and they've been eating that candy for years, but at some day you decide to take away the candy (make the public member private), then the child will cry an ocean of tears, and will eventually hate you.

As ridiculous as it may sound, but programmers do are similar to 5-year old children